### DISCUSSION TITLE
Renovate doesn't respect package-lock.json lockfileVersion to select NPM version to install

### DISCUSSION THREAD

**Discussion by kvanzuijlen** [renovatebot/renovate#25815] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.4

### If you're self-hosting Renovate, select which platform you are using.

Github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

    🚧 Doesn't work for both the GitHub App as for the self-hosted app

Previously, Renovate selected the NPM version to install based on configured constraints or the lockfileVersion it detected. This no longer works. I've set up a quick reproduction repo over here: https://github.com/kvanzuijlen/renovate-reproduction.

The issue this causes shows up in kvanzuijlen/renovate-reproduction#3

The error is as follows:
```log
ERROR: npm v10.2.3 is known not to run on Node.js v14.15.5.  This version of npm supports the following node versions: `^18.17.0 || >=20.5.0`. You can find the latest version at https://nodejs.org/.
```

The lockfileVersion is 1, so it should only be used with NPM v5 or v6 (see below), however, it tries to install NPM 10.2.3 (at the moment of writing latest) instead.

```
1: The lockfile version used by npm v5 and v6.
2: The lockfile version used by npm v7 and v8. Backwards compatible to v1 lockfiles.
3: The lockfile version used by npm v9. Backwards compatible to npm v7.
```

The node version is pinned to v14.15.5 by the .nvmrc file in the repository.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Generating package-lock.json for . (branch="renovate/express-4.x-lockfile")
DEBUG: Spawning npm install to create ./package-lock.json (branch="renovate/express-4.x-lockfile")
DEBUG: Updating lock file only (branch="renovate/express-4.x-lockfile")
DEBUG: Using node constraint "14.15.5" from .nvmrc (branch="renovate/express-4.x-lockfile")
DEBUG: Performing lockfileUpdate (npm) (branch="renovate/express-4.x-lockfile")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/worker/1619d6/0379b2/cache/containerbase (branch="renovate/express-4.x-lockfile")
DEBUG: Using docker to execute (branch="renovate/express-4.x-lockfile")
{
  "image": "sidecar"
}

DEBUG: Resolved stable matching version (branch="renovate/express-4.x-lockfile")
{
  "toolName": "npm"
  "constraint": null
  "resolvedVersion": "10.2.3"
}

```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Edit:** As far as I can tell this bug seems to have been introduced in #23131
**Edit 2:** I'm willing to work on a fix for this issue myself, but I'm currently at a loss on how I would go about fixing this

**Comment by rarkins** [comment]
You would want to add additional logic here: https://github.com/renovatebot/renovate/blob/da0a7e1037b43f5c7405e61bc53a56727048b0be/lib/modules/manager/npm/post-update/npm.ts#L47-L52

e.g. new function `getNpmConstraintFromPackageLock()` which is called if `getPackageManagerVersion()` does not return a value. I suggest changing `getPackageManagerVersion()` to return `undefined` instead of `null` so that you can then use `??` afterwards

**Reply by kvanzuijlen** [reply]
Thx! I'll take a look and I'll let you know!

**Reply by rarkins** [reply]
Fix PR: #25820

**Reply by fastman** [reply]
@rarkins Do you think this function should check `npm-shrinkwrap.json` if it exists? 

https://github.com/renovatebot/renovate/blob/37.133.1/lib/modules/manager/npm/post-update/npm.ts#L36

Maybe make that function take a parameter (lock/shrinkwrap) and call in:

https://github.com/renovatebot/renovate/blob/37.133.1/lib/modules/manager/npm/post-update/npm.ts#L80-L82

**Reply by rarkins** [reply]
Sure, PRs welcome

**Reply by fastman** [reply]
@rarkins  https://github.com/renovatebot/renovate/pull/26726

**Comment by rarkins** [comment, marked as answer]
Converted to bug issue: #25817
